### PR TITLE
Enable DPP and DPP+AQE on [databricks]

### LINKED ISSUE #3143: [BUG] DPP is not working in Databricks env
**Describe the bug**

DPP(dynamic partition pruning) is not working in Databricks env.
Found this issue when analyzing NDS query performance on Databricks.

**Steps/Code to reproduce bug**
Please provide a list of steps or a code sample to reproduce the issue.
Avoid posting private or sensitive data.

Below is the minimum reproduce:
```
import org.apache.spark.sql.functions.col
spark.range(1000).select(col("id"), col("id").as("k")).write.partitionBy("k").format("parquet").mode("overwrite").save("/tmp/hao/myfact")
spark.range(100).select(col("id"), col("id").as("k")).write.format("parquet").mode("overwrite").save("/tmp/hao/mydim")
spark.read.parquet("/tmp/hao/myfact").createOrReplaceTempView("fact")
spark.read.parquet("/tmp/hao/mydim").createOrReplaceTempView("dim")
spark.sql("SELECT fact.id, fact.k FROM fact JOIN dim ON fact.k = dim.k AND dim.id < 2").collect
```

GPU's physical plan:
```
Location: InMemoryFileIndex [dbfs:/tmp/hao/myfact]
PartitionFilters: [isnotnull(k#1265), dynamicpruningexpression(true)]
```

CPU's physical plan:
```
Location: InMemoryFileIndex [dbfs:/tmp/hao/myfact]
PartitionFilters: [isnotnull(k#2201), dynamicpruningexpression(cast(k#2201 as bigint) IN dynamicpruning#2211)]
```

As you can see, even though the dynamicpruningexpression keyword is there , however the filter is always true.
In Spark standalone cluster, there is no such issue for GPU run.

**Expected behavior**

DPP should happen in Databricks env.

**Environment details (please complete the following information)**
 - Environment location: [Standalone, YARN, Kubernetes, Cloud(specify cloud provider)]
 - Spark configuration settings related to the issue

Databricks runtime 8.2ML GPU
RAPIDS 21.10snapshot / 21.06GA

**Additional context**

WorkAround:
Disabling DPP by setting:
spark.sql.optimizer.dynamicPartitionPruning.enabled=false

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #3143.

This implements plugin support for Dynamic Partition Pruning (DPP) on Databricks distributions of Spark. 

Note that this enables DPP both AQE is enabled and disabled for both Databricks 9.1 and Databricks 10.4. This is a bit unusual since Databricks 9.1 is based on Apache Spark 3.1.2 (the open source version actually does not support both DPP+AQE being enabled at the same time), but the Databricks version does support this cooperation in the closed source distribution.  In fact, both DPP and AQE are enabled by default in Databricks 9.1, so this is done to support the default configuration.

### CODE REVIEW

**Comment by NVnavkumar** [comment]
build

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala [hunk 1]
```diff
@@ -284,7 +284,7 @@ class GpuBroadcastMeta(
     conf: RapidsConf,
     parent: Option[RapidsMeta[_, _, _]],
     rule: DataFromReplacementRule) extends
-  SparkPlanMeta[BroadcastExchangeExec](exchange, conf, parent, rule) {
+  SparkPlanMeta[BroadcastExchangeExec](exchange, conf, parent, rule) with Logging {
```
Comment (jlowe) [inline sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala:287]: Change seems unnecessary?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastHashJoinExec.scala [hunk 2]
```diff
@@ -147,6 +147,9 @@ case class GpuBroadcastHashJoinExec(
       bqse.plan.asInstanceOf[ReusedExchangeExec].child.asInstanceOf[GpuBroadcastExchangeExec]
     case gpu: GpuBroadcastExchangeExec => gpu
     case reused: ReusedExchangeExec => reused.child.asInstanceOf[GpuBroadcastExchangeExec]
+    case _ => 
+      logWarning(s"what is this? ${buildPlan.getClass}")
```
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastHashJoinExec.scala:151]: Oops, this was debug code, it should be removed.

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala [hunk 3]
```diff
@@ -55,6 +59,33 @@ abstract class Spark31XdbShims extends Spark31XdbShimsBase with Logging {
...
+      )
+      rules.foldLeft(plan) { case (sp, rule) =>
+        rule.apply(sp)
+      }
+    }
+  }
+
+  override def applyPostShimPlanRules(plan: SparkPlan): SparkPlan = {
+    if (plan.conf.adaptiveExecutionEnabled) {
+      plan // no need to handle DPP here since it doesn't cooperate with AQE
```
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala:78]: yes, this comment was out of date, will correct

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala [hunk 4]
```diff
@@ -155,6 +186,100 @@ abstract class Spark31XdbShims extends Spark31XdbShimsBase with Logging {
...
+            case bqse: BroadcastQueryStageExec =>
+              bqse.plan match {
+                case ex: BroadcastExchangeExec =>
+                  val exMeta = new GpuBroadcastMeta(ex, conf, p, r)
+                  exMeta.tagForGpu()
+                  if (exMeta.canThisBeReplaced) {
+                    broadcastBuilder = () => exMeta.convertToGpu()
+                    // broadcastBuilder = () =>
+                    //   SparkShimImpl.columnarAdaptivePlan(
+                    //     a, TargetSize(conf.gpuTargetBatchSizeBytes))
```
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala:230]: No, this should be removed in both shims

File: sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala [hunk 5]
```diff
@@ -80,8 +84,125 @@ object SparkShimImpl extends Spark321PlusShims with Spark320until340Shims {
...
+            case bqse: BroadcastQueryStageExec =>
+              bqse.plan match {
+                case ex: BroadcastExchangeExec =>
+                  val exMeta = new GpuBroadcastMeta(ex, conf, p, r)
+                  exMeta.tagForGpu()
+                  if (exMeta.canThisBeReplaced) {
+                    broadcastBuilder = () => exMeta.convertToGpu()
+                    // broadcastBuilder = () =>
+                    //   SparkShimImpl.columnarAdaptivePlan(
+                    //     a, TargetSize(conf.gpuTargetBatchSizeBytes))
```
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply sql-plugin/src/main/321db/scala/com/nvidia/spark/rapids/shims/SparkShims.scala]: removed

File: sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala (hunk 1 above)
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuBroadcastExchangeExec.scala]: removed

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastHashJoinExec.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuBroadcastHashJoinExec.scala]: removed

**Comment by NVnavkumar** [comment]
build

File: sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala [hunk 6]
```diff
@@ -16,7 +16,9 @@
 
 package com.nvidia.spark.rapids.shims
 
+// import com.databricks.sql.execution.ReuseExchangeAndSubquery
```
Comment (jlowe) [inline sql-plugin/src/main/31xdb/scala/com/nvidia/spark/rapids/shims/Spark31XdbShims.scala:19]:
```suggestion
```

File: sql-plugin/src/main/31xdb/scala/org/apache/spark/sql/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+  override val childExprs: Seq[BaseExprMeta[_]] = Nil
+
+  override val childPlans: Seq[SparkPlanMeta[SparkPlan]] = Nil
+
+  override def tagPlanForGpu(): Unit = s.child match {
+    // DPP: For AQE off, in this case, we handle DPP by converting the underlying 
+    // BroadcastExchangeExec to GpuBroadcastExchangeExec
+    // This is slightly different from the Apache Spark case, because Spark 
+    // sends the underlying plan into the plugin in advance via the PlanSubqueries rule
```
Comment (jlowe) [inline sql-plugin/src/main/31xdb/scala/org/apache/spark/sql/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala:46]:
```suggestion
    // sends the underlying plan into the plugin in advance via the PlanSubqueries rule.
```

File: sql-plugin/src/main/31xdb/scala/org/apache/spark/sql/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+
+  private var broadcastBuilder: () => SparkPlan = _
+
+  override val childExprs: Seq[BaseExprMeta[_]] = Nil
+
+  override val childPlans: Seq[SparkPlanMeta[SparkPlan]] = Nil
+
+  override def tagPlanForGpu(): Unit = s.child match {
+    // DPP: For AQE off, in this case, we handle DPP by converting the underlying 
+    // BroadcastExchangeExec to GpuBroadcastExchangeExec
```
Comment (jlowe) [inline sql-plugin/src/main/31xdb/scala/org/apache/spark/sql/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala:44]:
```suggestion
    // BroadcastExchangeExec to GpuBroadcastExchangeExec.
```

File: sql-plugin/src/main/321db/scala/org/apache/spark/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+    SparkPlanMeta[SubqueryBroadcastExec](s, conf, p, r) {
+  private var broadcastBuilder: () => SparkPlan = _
+
+  override val childExprs: Seq[BaseExprMeta[_]] = Nil
+
+  override val childPlans: Seq[SparkPlanMeta[SparkPlan]] = Nil
+
+  override def tagPlanForGpu(): Unit = s.child match {
+    // DPP: For AQE off, in this case, we handle DPP by converting the underlying 
+    // BroadcastExchangeExec to GpuBroadcastExchangeExec
```
Comment (jlowe) [inline sql-plugin/src/main/321db/scala/org/apache/spark/rapids/execution/shims/GpuSubqueryBroadcastMeta.scala:43]:
```suggestion
    // BroadcastExchangeExec to GpuBroadcastExchangeExec.
```

**Review by jlowe — COMMENTED** [review]
Minor nits but otherwise lgtm.

**Review by jlowe — APPROVED** [review]

**Pull request merged**